### ISSUE TITLE
Need better parsing of serial command

### ISSUE THREAD

**Issue by scottmc** [TeamNorCal/arduino-resonators#5]
need to verify that the serial command fits the spec and if not discard it and get the next

must begin with:
e,E,r,R,n or N
next 8 chars must be between 0 and 8, no exceptions.

Characters after that should determine resonator percentage per string.  This doesn't seem to be working correct yet.

**Comment by aroneous** [comment]
It does seem like writing to the LEDs (probably involving disabling of interrupts) is the cause of this flakiness. Commenting out the call to doFrame (which is where the LED updates occur) makes serial reception 100% reliable. With the doFrame call present, it misses at least one character around 1/3 of the time.

I tried lowering the baud rate from 115200 -> 2400, which improved reliability substantially - I saw one missed character out of about 20 commands. That would be consistent with disabled interrupts being a cause of the problem, as there is more time to process each character at lower baud rates

The problem will be worse with longer strings of LEDs, as all of them are written at once so the time interrupts are disabled is proportional to number of LEDs. The numbers quoted above were using the full 240 LED strips. Moving to 108 LED strips (which is what Scott indicated will be used in practice) did improve performance somewhat - I experienced about 20% failure rate on 20 tries at 115200 baud, and actually somewhat higher failure rate at 9600 baud, surprisingly. I doubt the difference is statistically significant, but I would have expected noticeably better behavior at 9600 baud.

I'll study the NeoPixel code a bit to see if anything can be improved there, but our best solution might be to lower baud rate to 2400 or so (which I'd expect to be fine given the limited amount of data being transmitted) and adding an application-level checksum, or perhaps insisting on fixed-length commands, discarding any that aren't the expected length. If we go with fixed-length commands, that will imply a change to controller code.

**Comment by aroneous** [comment]
Good wiki post [here](https://github.com/FastLED/FastLED/wiki/Interrupt-problems) discussing this very problem (which is, indeed, interrupt-related). Possible solutions:

1. Throw hardware at it - moving to an ARM-based Arduino like the Due provides enough grunt to briefly enable interrupts between each pixel, so characters can be serviced
2. Implement a pull-based protocol, where Arduino sends a request for data at a convenient point, and the controller responds
3. When serial input arrives, stop servicing LEDs until the command is complete. (Currently we read what's in the serial buffer and, if the command isn't complete, we service the LEDs and come back to look for more input. This change implies we'd stop servicing the LEDs until we read the entirety of the command.) In this case, they suggest looking for a series of marker bytes at the start of the command so you know you didn't miss initial bytes; for us, that's probably handled implicitly, since if the first byte isn't one or eErRn*, we ignore it anyway, and those can't exist elsewhere in the command string.

Based on all this, I'm planning to move to option #3, as the simplest way to deal with this. It effectively serves the purpose of a checksum and should be more reliable than a checksum where we continue to service the LEDs while command text is being transmitted.

**Comment by aroneous** [comment]
Initial attempts at the strategy of not servicing LEDs while serial input is arriving have been disappointing. Debug logging shows that when we get into collect_serial, the full command is always already waiting in the serial buffer, so the dropped-byte problem has already happened before our code has any awareness that input is arriving.

If I'm giving a 4-byte command like "r12\n", I've only ever seen it miss the '2' (most commonly) or the '\n'. I've never seen it miss one of the first two bytes. Would the NeoPixel code limit the frequency of the interrupt processing, such that interrupts run for the first couple of characters, but then it stops allowing them such that a character is missed?

Trying with longer (nonsensical) command 'r123456789\n', the failure modes I've seen include:

- r1\n (most of the time)
- r123456789 (missing the newline - this portion came through along with the next command, read successfully)
- r1789\n (missing 23456, interestingly)

The last two I saw once each after around 20 tries; I saw the first failure mode maybe 5 times.
